### DISCUSSION TITLE
How can the receiver of a RNS message authenticate the sender of it?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#315] [Help & Questions] [answered]
I need to understand how to authenticate who the sender of a message is. 

If a sender creates an identity, saves it to a file, and always uses that file to create destinations, can the hash of that identity be used as a a config parameter of receivers to authenticate the sender of messages from all destinations created with that identity?

Assuming the sender uses a link with identity enabled to communicate.

**Comment by markqvist** [comment, marked as answer]
Short answer: When the sender of the message is using a `Link` and uses the `identify()` API call. This functionality is directly built into the protocol and API.

First, you have to realise that all communication in Reticulum is *anonymous by default*. No packets or link establishments carry any identifying information from the outset. **All packets** are sent without any type of source addresses or identifiers.

In Reticulum, any `Destination` that you *send something to* is always authenticated by default and automatically. This is one of the primary design points of Reticulum. If you send a packet to a `Destination` in Reticulum, you can be sure that only that particular `Destination`, identified by it's 16-byte hash, will be able to receive and decrypt it.

Understanding the above, you can see how Reticulum destinations ("addresses") are directly useful as authenticated endpoints (and should be used as such). This is very different from IP, where anyone and anything could be behind a given IP address.

Take as an example how the LXMF messaging system directly uses Reticulum destination hashes as user addresses:

- An LXMF client generates a new Reticulum `Identity`, and saves it to user storage.
- It can now use this `Identity` to create a Reticulum `Destination` with the aspects `lxmf.delivery`.
- The hash of this `Destination` is directly the users LXMF address.
- The user now has a globally reachable address, that is portable throughout any Reticulum networks in existence.
- This address can now be used by anyone else to send messages to the user securely.

The above is (one of) the right ways to do things. When you want addresses that are tied to some sort of longer-term trust or recognisability, save the `Identity` and load it, and use it to create your destinations. It's a complete identity construct, but self-sovereign, portable, and without requiring any external authority. Neat, right?

In some situations, you may want to create completely ephemeral destinations. Just create a new `Identity`, create your destinations from that, announce them, do whatever you want, and discard the `Identity` when you are done. This is how all the Reticulum code examples work, and demonstrates how you can create as many ephemeral addresses as you want. In most *real world applications* though, this is probably **not** what you want to do.

**Think** about what is right for your application. **Don't** try to do things like you do with TCP/IP, where you have to plaster all these extra layers of security and authentication *on top of the base-layer*. In Reticulum, it is already there, you just have to use them in the right way. Your application will be much simpler and orders of magnitude more bandwidth-efficient.

Initiators (senders) *can* identify towards a remote destination, **if they choose to**, but these points are essential to remember:

- Cryptographic identification (and therefore authentication) can only happen over an established `Link`, such that the identifcation process is happening with forward secrecy, and nobody else than the intended `Destination` can see the identification.
- Only the initiator of the link has the authority to identify towards the remote, the listener cannot request the initiator to do so. I repeat, the initiator of the link must *actively* choose to identify towards the destination.
- Carrying out an identification only costs one packet over the link, and will unambigiously authenticate the link initiator for the listener.
- If an initiator carries out an `identify()` call, both ends of the link are now authenticated - remember that the destination is always authenticated by default.

**Reply by ThomasFreedman** [reply]
Thanks for the clarification. Some of that I understood, but not everything, such as the last point. This really does simplify my app regarding the identities of both parties.

**Reply by markqvist** [reply]
Read it all a few more times, it will make sense. The listening end of the link was already authenticated when you established the link ;)
